core/src/main/scala/org/apache/spark/scheduler/TaskSchedulerImpl.scala

```diff
@@ -939,12 +941,43 @@ private[spark] class TaskSchedulerImpl(
 
   override def executorDecommission(
       executorId: String, decommissionInfo: ExecutorDecommissionInfo): Unit = {
+    synchronized {
+      // Don't bother noting decommissioning for executors that we don't know about
+      if (executorIdToHost.contains(executorId)) {
+        // The scheduler can get multiple decommission updates from multiple sources,
+        // and some of those can have isHostDecommissioned false. We merge them such that
+        // if we heard isHostDecommissioned ever true, then we keep that one since it is
+        // most likely coming from the cluster manager and thus authoritative
+        val oldDecomInfo = executorsPendingDecommission.get(executorId)
+        if (oldDecomInfo.isEmpty || !oldDecomInfo.get.isHostDecommissioned) {
+          executorsPendingDecommission(executorId) = decommissionInfo
+        }
+      }
+    }
     rootPool.executorDecommission(executorId)
     backend.reviveOffers()
   }
 
-  override def executorLost(executorId: String, reason: ExecutorLossReason): Unit = {
+  override def getExecutorDecommissionInfo(executorId: String)
+    : Option[ExecutorDecommissionInfo] = synchronized {
+      executorsPendingDecommission.get(executorId)
+  }
+
+  override def executorLost(executorId: String, givenReason: ExecutorLossReason): Unit = {
     var failedExecutor: Option[String] = None
+    val reason = givenReason match {
+      // Handle executor process loss due to decommissioning
+      case ExecutorProcessLost(message, workerLost, causedByApp) =>
+        val executorDecommissionInfo = getExecutorDecommissionInfo(executorId)
+        ExecutorProcessLost(
+          message,
+          // Also mark the worker lost if we know that the host was decommissioned
+          workerLost || executorDecommissionInfo.exists(_.isHostDecommissioned),
+          // Executor loss is certainly not caused by app if we knew that this executor is being
+          // decommissioned
+          causedByApp = executorDecommissionInfo.isEmpty && causedByApp)
```

Comment:
We add the `causedByApp` parameter, because we may decommission an executor but not a host?

I'm thinking if `!workerLost` can indicate that the failure is caused by the app.

Reply:
As the documentation of the class `ExecutorProcessLost` says:

`causedByApp` is used to track whether to penalize the failure of the executor to the task or not. ie, should we attribute the failure of the executor to the usual limit of 4 failures of a task or not. This is useful for example in certain YARN like scenarios, where a single task can blow up the entire executor's memory budget and thus be killed by YARN. In this case, we do want to retain the default behavior of `causedByApp = true`. The default spark behavior has been to _always_ penalize the task for all executor losses (heartbeat failures) and I didn't want to change that..

`workerLost` is to track whether the shuffle files resident in the external shuffle service (which is local to a worker and not to a single executor) have been lost or not. Recall that worker == node :-). So `workerLost` is used to clear the shuffle state of all the executors on the node to avoid subsequent fetch failures. 

So I would rather keep these two flags separate since they are for tracking two orthogonal things. Indeed there can be these cases when `causedByApp == workerLost`:

- `causedByApp = true, workerLost = true`: Currently only happens when the standalone (master) scheduler removes an executor without an exit code
- `causedByApp = false, workerLost = false`: As you pointed out, this can happen when an executor is decommissioned without the host in K8s mode.